src/components/templates/TalkShowStudio/useStage.ts

```diff
@@ -0,0 +1,194 @@
+    };
+  }, [setDefaultUserSettings]);
+
+  const peopleOnStage = useMemo(
+    () =>
+      venueId
+        ? recentVenueUsers.filter(
+            (user) =>
+              user.data?.[venueId]?.place === PlaceInTalkShowStudioVenue.stage
+          )
+        : [],
+    [recentVenueUsers, venueId]
+  );
+
+  const peopleRequesting = useMemo(
+    () =>
+      venueId
+        ? recentVenueUsers.filter(
+            (user) =>
+              user.data?.[venueId]?.place ===
+              PlaceInTalkShowStudioVenue.requesting
+          )
+        : [],
+    [recentVenueUsers, venueId]
+  );
+
+  const isUserRequesting = useMemo(
+    () => peopleRequesting.some((user) => user.id === userId),
+    [peopleRequesting, userId]
+  );
+
+  const isUserOnStage = useMemo(
+    () => peopleOnStage.some((user) => user.id === userId),
+    [peopleOnStage, userId]
+  );
+
+  const canJoinStage = useMemo(
+    () => peopleOnStage.length < MAX_TALK_SHOW_STUDIO_HOSTS,
+    [peopleOnStage]
+  );
+
+  const screenSharingUser = useMemo(
+    () =>
+      peopleOnStage.find(
+        (user) => venueId && user.data?.[venueId].isSharingScreen
+      ),
+    [peopleOnStage, venueId]
+  );
+
+  const canShareScreen = useMemo(() => !screenSharingUser, [screenSharingUser]);
+
+  const isUserSharing = useMemo(
+    () => !!userId && screenSharingUser?.id === userId,
+    [screenSharingUser, userId]
+  );
+
+  const joinStage = () => {
+    venueId &&
+      userId &&
+      updateTalkShowStudioExperience({
+        venueId,
+        userId,
+        experience: {
+          place: PlaceInTalkShowStudioVenue.stage,
+        },
+      });
+  };
+
+  const leaveStage = async () => {
+    // TODO: clear agoraIds for user when left Stage
+    await setDefaultUserSettings();
+  };
+
+  const requestJoinStage = () => {
+    venueId &&
+      userId &&
+      updateTalkShowStudioExperience({
+        venueId,
+        userId,
+        experience: {
+          place: PlaceInTalkShowStudioVenue.requesting,
+        },
+      });
+  };
+
+  const toggleMicrophone = () => {
+    venueId &&
+      userId &&
+      updateTalkShowStudioExperience({
+        venueId,
+        userId,
+        experience: {
+          isMuted: !profile?.data?.[venueId].isMuted,
+        },
+      });
+  };
+
+  const toggleCamera = () => {
+    venueId &&
+      userId &&
+      updateTalkShowStudioExperience({
+        venueId,
+        userId,
+        experience: {
+          isUserCameraOff: !profile?.data?.[venueId].isUserCameraOff,
+        },
+      });
+  };
+
+  const toggleUserMicrophone = (user?: WithId<User>) => {
+    if (user && venueId) {
+      updateUserTalkShowStudioExperience(venueId, user.id, {
+        isMuted: true,
+      });
+    }
+  };
+
+  const toggleUserCamera = (user?: WithId<User>) => {
+    if (user && venueId) {
+      updateUserTalkShowStudioExperience(venueId, user.id, {
+        isUserCameraOff: true,
+      });
+    }
+  };
```

Comment:
Why do these need to use `updateUserTalkShowStudioExperience` / be persisted in firebase? Are these not things we can read directly out of the Agora state? That's how we do it with Twilio I believe.